nmigen_boards/blackice.py

```diff
@@ -0,0 +1,80 @@
+import os
+import subprocess
+
+from nmigen.build import *
+from nmigen.vendor.lattice_ice40 import *
+
+
+__all__ = ["BlackIcePlatform"]
+
+
+class BlackIcePlatform(LatticeICE40Platform):
+    device = "iCE40HX8K"
+    package = "TQ144:4K"
+    resources  = [
+        Resource("clk100", 0, Pins("129", dir="i"),
+            Clock(100e6), Attrs(GLOBAL="1", IO_STANDARD="SB_LVCMOS33")
+        ),
+
+        Resource("user_led", 0, Pins("71", dir="o"), Attrs(IO_STANDARD="SB_LVCMOS33")),
+        Resource("user_led", 1, Pins("67", dir="o"), Attrs(IO_STANDARD="SB_LVCMOS33")),
+        Resource("user_led", 2, Pins("68", dir="o"), Attrs(IO_STANDARD="SB_LVCMOS33")),
+        Resource("user_led", 3, Pins("70", dir="o"), Attrs(IO_STANDARD="SB_LVCMOS33")),
+        # Color aliases
+        Resource("user_ledb", 0, Pins("71", dir="o"), Attrs(IO_STANDARD="SB_LVCMOS33")),
+        Resource("user_ledg", 0, Pins("67", dir="o"), Attrs(IO_STANDARD="SB_LVCMOS33")),
+        Resource("user_ledo", 0, Pins("68", dir="o"), Attrs(IO_STANDARD="SB_LVCMOS33")),
+        Resource("user_ledr", 0, Pins("70", dir="o"), Attrs(IO_STANDARD="SB_LVCMOS33")),
+
+        Resource("user_btn", 0, PinsN("63", dir="i"), Attrs(IO_STANDARD="SB_LVCMOS33")),
+        Resource("user_btn", 1, PinsN("64", dir="i"), Attrs(IO_STANDARD="SB_LVCMOS33")),
+
+        Resource("user_sw", 0, PinsN("37", dir="i"), Attrs(IO_STANDARD="SB_LVCMOS33")),
+        Resource("user_sw", 1, PinsN("38", dir="i"), Attrs(IO_STANDARD="SB_LVCMOS33")),
+        Resource("user_sw", 2, PinsN("39", dir="i"), Attrs(IO_STANDARD="SB_LVCMOS33")),
+        Resource("user_sw", 3, PinsN("41", dir="i"), Attrs(IO_STANDARD="SB_LVCMOS33")),
+
+        Resource("serial", 0,
+            Subsignal("rx", Pins("88", dir="i")),
+            Subsignal("tx", Pins("85", dir="o")),
+            Subsignal("rts", Pins("91", dir="o")),
+            Subsignal("cts", Pins("94", dir="i")),
+            Attrs(IO_STANDARD="SB_LVCMOS33"),
+        ),
+
+        Resource("sram", 0,
+            Subsignal("address", Pins(
+                "137 138 139 141 142 42 43 44 73 74 75 76 115 116 117 118 119 78",
+                dir="o"
+            )),
+            Subsignal("data", Pins(
+                "135 134 130 128 125 124 122 121 61 60 56 55 52 49 48 47",
+                dir="io"
+            )),
+            Subsignal("oe", PinsN("45", dir="o")),
+            Subsignal("we", PinsN("120", dir="o")),
+            Subsignal("cs", PinsN("136", dir="o")),
+            Attrs(IO_STANDARD="SB_LVCMOS33"),
+        ),
+    ]
+
+    connectors = [ # TODO: check order of top/bottom row
```

Comment:
Is this still a TODO? I don't understand what it says anyway.

Reply:
No, forgot to remove comment for BlackIce.
Based file on migen boards file and there the double PMOD connectors were split in two. I joined them but made note to check if top row and bottom row matched. As I used the PMOD connectors now the correctness has been verified.
Solved in latest push.